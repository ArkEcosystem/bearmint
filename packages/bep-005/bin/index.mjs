#!/usr/bin/env node
import { join } from 'node:path'

import { globby } from 'globby'
import meow from 'meow'

process.env.NODE_ENV = 'test'

const { flags } = meow({
	flags: {
		config: {
			type: 'string',
		},
	},
	importMeta: import.meta,
})

const { files } = (await import(join(process.cwd(), flags.config))).default
const tests = (await globby(files)).map((path) => join(process.cwd(), path))

for (const test of tests) {
	await import(test)
}
