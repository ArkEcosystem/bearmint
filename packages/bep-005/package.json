{
	"name": "@bearmint/bep-005",
	"version": "0.0.0",
	"private": true,
	"description": "This library is the official reference implementation for BEP-005 (https://bep.dev/005)",
	"keywords": [
		"bearmint",
		"tendermint",
		"blockchain",
		"framework",
		"modular"
	],
	"homepage": "https://github.com/bearmint/bearmint",
	"repository": {
		"type": "git",
		"url": "git+https://github.com/bearmint/bearmint.git"
	},
	"funding": "https://funding.bearmint.com/",
	"license": "MIT",
	"author": "Bearmint Mining Company <hello@bearmint.com> (https://bearmint.com/)",
	"type": "module",
	"exports": "./distribution/index.js",
	"types": "./distribution/index.d.ts",
	"bin": {
		"roadrunner": "./bin/index.mjs"
	},
	"files": [
		"bin",
		"distribution"
	],
	"scripts": {
		"audit": "pnpm audit --fix --prod",
		"build": "pnpm run clean && tsc",
		"build:watch": "pnpm run clean && tsc -w",
		"clean": "rm -rf distribution",
		"depcheck": "depcheck --ignores=@bearmint/bep-004,@bearmint/bep-005,@bearmint/bep-031,@bearmint/bep-039,globby,meow",
		"deps:major": "ncu -u --target=latest",
		"deps:minor": "ncu -u --target=minor",
		"deps:patch": "ncu -u --target=patch"
	},
	"dependencies": {
		"@bearmint/bep-013": "workspace:^",
		"@bearmint/bep-018": "workspace:^",
		"@bearmint/bep-116": "workspace:^",
		"@sinonjs/fake-timers": "9.1.2",
		"ansi-styles": "6.2.1",
		"concordance": "5.0.4",
		"globby": "13.1.2",
		"lodash.get": "4.4.2",
		"lodash.has": "4.5.2",
		"meow": "^11.0.0",
		"nock": "13.2.9",
		"pretty-ms": "8.0.0",
		"sinon": "14.0.1",
		"strip-ansi": "7.0.1",
		"vitest": "^0.24.5"
	},
	"devDependencies": {
		"@types/node": "18.11.9",
		"@types/sinon": "10.0.13",
		"@types/sinonjs__fake-timers": "8.1.2"
	},
	"engines": {
		"node": ">=18",
		"pnpm": ">=7.9.0"
	},
	"publishConfig": {
		"access": "restricted"
	}
}
