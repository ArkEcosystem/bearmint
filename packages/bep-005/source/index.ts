export * from './expect.js'
export * from './fake.js'
export * from './nock.js'
export * from './spy.js'
export * from './stub.js'
