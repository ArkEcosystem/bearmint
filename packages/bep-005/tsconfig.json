{
	"compilerOptions": {
		"typeRoots": ["node_modules/@types"],
		"outDir": "distribution"
	},
	"exclude": ["node_modules", "test/**/**.ts"],
	"extends": "../../config/tsconfig.json",
	"include": ["source/**/**.ts"]
}
