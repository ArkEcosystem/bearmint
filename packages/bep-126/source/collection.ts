import type { Account, KVStore, StateStore, TxAssertionsErrorStruct } from '@bearmint/bep-013'
import { PROTO_OPTS } from '@bearmint/bep-013'
import type { Plain } from '@bearmint/bep-018'
import { BEP126Collection, BEP131Token } from '@bearmint/bep-018'
import { TxAuditException } from '@bearmint/bep-109'

import { BEP126UserPermission, HANDLER } from './types.js'

export async function getCollection(bep126Store: KVStore, hash: string) {
	return BEP126Collection.fromBinary(await bep126Store.get(hash), PROTO_OPTS)
}

export async function getCollectionHash(bep131Store: KVStore, hash: string) {
	return BEP131Token.fromBinary(await bep131Store.get(hash), PROTO_OPTS).collection
}

export async function getCollectionFromToken(
	bep126Store: KVStore,
	bep131Store: KVStore,
	hash: string,
) {
	// 1. Retrieve the hash of the collection the NFT belongs to
	const collectionHash = await getCollectionHash(bep131Store, hash)

	// 2. Retrieve the actual collection
	const collection = await getCollection(bep126Store, collectionHash)

	// 3. Return both to allow for easy referencing
	return { collection, collectionHash }
}

export async function hasCollection(bep126Store: KVStore, hash: string) {
	return bep126Store.has(hash)
}

export async function hasCollectionOwnership(state: StateStore, hash: string, account: Account) {
	try {
		return (
			(await state.getMultiStore().get(HANDLER).getListHex(account.address)).includes(hash) === true
		)
	} catch {
		return false
	}
}

export async function assertCollectionOwnership(
	state: StateStore,
	hash: string,
	account: Account,
	error: TxAssertionsErrorStruct,
) {
	if (await hasCollectionOwnership(state, hash, account)) {
		return
	}

	throw new TxAuditException(error)
}

export function isUnknownPermission(permission: number, error: TxAssertionsErrorStruct) {
	if (permission in BEP126UserPermission) {
		return
	}

	throw new TxAuditException(error)
}

export async function upsertCollection(
	bep126Store: KVStore,
	collectionHash: string,
	collection: Plain<BEP126Collection>,
) {
	const result = new BEP126Collection(collection)

	await bep126Store.set(collectionHash, result.toBinary())

	return result
}
