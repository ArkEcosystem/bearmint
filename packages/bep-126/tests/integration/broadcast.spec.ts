import { expect, describe, it, beforeEach } from 'vitest'
import { extendExpect } from '@bearmint/bep-005'
extendExpect(expect)
import type { TxTestContext } from '@bearmint/bep-006'

import { getCollection } from '../../source/collection.js'
import { STUB_TX_META } from '../fixtures.js'
import { createTx, setUp } from '../setup.js'

describe<TxTestContext>('Tx Integration', () => {
	setUp(beforeEach)

	it('should run through the entire lifecycle of processing the tx', async (context) => {
		const { sender, tx } = await createTx(context)

		expect(
			await context.state.getMultiStore().get('@bearmint/bep-126').has(STUB_TX_META.id),
		).toBeFalse()

		await context.executeFull(tx)

		expect(
			await context.state.getMultiStore().get('@bearmint/bep-126').getListHex(sender.address),
		).toStrictEqual([STUB_TX_META.id])

		expect(
			await getCollection(context.state.getMultiStore().get('@bearmint/bep-126'), STUB_TX_META.id),
		).toMatchSnapshot()
	})
})
