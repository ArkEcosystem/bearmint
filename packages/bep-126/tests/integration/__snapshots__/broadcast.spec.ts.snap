// Vitest Snapshot v1

exports[`Tx Integration > should run through the entire lifecycle of processing the tx 1`] = `
BEP126Collection {
  "accessControlList": {
    "bear15z20cwcgszr54v7w7pxulyv3pyqjz878rmvzqtjf2sqphz7gv4hrz0f44vuwgx57qy6h0c064x3rw44zxpl": BEP126User {
      "permissions": [
        0,
        1,
        2,
      ],
      "role": 0,
    },
  },
  "currentOwner": "bear15z20cwcgszr54v7w7pxulyv3pyqjz878rmvzqtjf2sqphz7gv4hrz0f44vuwgx57qy6h0c064x3rw44zxpl",
  "id": "e2abe831494bd6d8642810cb11e2d36944abb2206c40ac5f24be73d92258dfa9",
  "initialOwner": "bear15z20cwcgszr54v7w7pxulyv3pyqjz878rmvzqtjf2sqphz7gv4hrz0f44vuwgx57qy6h0c064x3rw44zxpl",
  "name": "name",
  "policies": [
    BEP126Policy {
      "args": "",
      "name": "@bearmint/bep-145",
      "type": 0,
    },
    BEP126Policy {
      "args": "",
      "name": "@bearmint/bep-145",
      "type": 1,
    },
    BEP126Policy {
      "args": "",
      "name": "@bearmint/bep-145",
      "type": 2,
    },
  ],
  "symbol": "symbol",
  "tokens": [],
  "uriPrefix": "uriPrefix",
  "uriSuffix": "uriSuffix",
}
`;
